### PR TITLE
[FEAT] Add `horizon_weight` parameter to losses and `BasePointLoss`

### PR DESCRIPTION
**Pull request by cchallu**
* Added `horizon_weight` parameter to all point and MQ losses. Used to weight different timestamps of the forecasting window.
* New `BasePointLoss` with common methods.
* Modified all point and MQ losses to inherit from `BasePointLoss` class.

### CODE REVIEW

File: neuralforecast/losses/pytorch.py [hunk 1]
```diff
@@ -635,20 +659,20 @@ def __call__(
...
-            torch.sum(loss * mask, axis=-2),
-            torch.sum(torch.abs(y.unsqueeze(-1)) * mask, axis=-2),
-        )
-        return torch.mean(wmqloss)
+        mqloss = (1 / n_q) * mqloss * mask
+        return torch.sum(mqloss)
 
-# %% ../../nbs/losses.pytorch.ipynb 60
+# %% ../../nbs/losses.pytorch.ipynb 59
 def weighted_average(
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:676]: Can we rescue something from this `weighted_average` function?
Or will we deprecate it?

File: neuralforecast/losses/pytorch.py [hunk 2]
```diff
@@ -560,32 +575,41 @@ def __call__(
...
+        weights = self._compute_weights(y=losses, mask=mask)  # Use losses for extra dim
+        # NOTE: Weights do not have Q dimension.
+
+        return _weighted_mean(losses=losses, weights=weights)
 
-# %% ../../nbs/losses.pytorch.ipynb 55
-class wMQLoss(torch.nn.Module):
-    """Weighted Multi-Quantile loss
+# %% ../../nbs/losses.pytorch.ipynb 53
+class MQLoss_deprecated(torch.nn.Module):
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:595]: Let's just erase this MQLoss_deprecated class, it is redundant with sCRPS

File: neuralforecast/losses/pytorch.py [hunk 3]
```diff
@@ -27,24 +28,31 @@ def _divide_no_nan(a: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
     div[div == float("inf")] = 0.0
     return div
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:29]: If we are taking numerical protections in the denominator of certain losses, lets add the option for the numerical protection to _divide_no_nan as an alternative. See comments on relMSE.

File: neuralforecast/losses/pytorch.py [hunk 4]
```diff
@@ -280,20 +294,16 @@ def __call__(
         `mask`: tensor, Specifies date stamps per serie to consider in loss.<br>
 
         **Returns:**<br>
-        `smape`: tensor (single value).
+        `mape`: tensor (single value).
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:297]: you are missing an s in `smape`

File: neuralforecast/losses/pytorch.py [hunk 5]
```diff
@@ -391,27 +394,25 @@ def __call__(
...
-        n_series, horizon = y.shape
-
+        horizon = y.shape[-1]
         last_col = self.y_train[:, -1].unsqueeze(1)
         y_naive = last_col.repeat(1, horizon)
 
-        norm = self.mse(y=y, y_hat=y_naive)
-        loss = self.mse(y=y, y_hat=y_hat, mask=mask)
-        loss = loss / (norm + 1e-5)
+        norm = self.mse(y=y, y_hat=y_naive, mask=mask)  # Already weighted
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:409]: Recover the numerical protection on norm.

File: neuralforecast/losses/pytorch.py [hunk 6]
```diff
@@ -560,32 +575,41 @@ def __call__(
...
+        losses = (1 / len(self.quantiles)) * (
+            self.quantiles * sq + (1 - self.quantiles) * s1_q
+        )
 
-        # Match y/weights dimensions and compute weighted average
-        mask = mask / torch.sum(mask)
-        mask = mask.unsqueeze(-1)
-        mqloss = (1 / n_q) * mqloss * mask
-        return torch.sum(mqloss)
+        losses = losses.swapaxes(
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:586]: The swapaxes solution seems to overlap with the previous weighted_mean used in DistributionLoss

File: neuralforecast/losses/pytorch.py [hunk 7]
```diff
@@ -1830,15 +1847,13 @@ def __call__(
         **Returns:**<br>
         `huber_loss`: tensor (single value).
         """
-        if mask is None:
-            mask = torch.ones_like(y_hat)
+        losses = F.huber_loss(y, y_hat, reduction="none", delta=self.delta)
+        print(losses)
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:1851]: forgot print

File: neuralforecast/losses/pytorch.py [hunk 8]
```diff
@@ -1830,15 +1847,13 @@ def __call__(
...
-        huber_loss = F.huber_loss(
-            y * mask, y_hat * mask, reduction="mean", delta=self.delta
-        )
-        return huber_loss
+        weights = self._compute_weights(y=y, mask=mask)
+        return _weighted_mean(losses=losses, weights=weights)
 
-# %% ../../nbs/losses.pytorch.ipynb 97
+# %% ../../nbs/losses.pytorch.ipynb 96
 class TukeyLoss(torch.nn.Module):
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:1857]: Why is TukeyLoss not converted like the rest of the point losses?

File: neuralforecast/losses/pytorch.py [hunk 9]
```diff
@@ -2012,34 +2021,59 @@ class HuberMQLoss(torch.nn.Module):
...
-        qs, self.output_names = level_to_outputs(level)
+    def __init__(
+        self, level=[80, 90], quantiles=None, delta: float = 1.0, horizon_weight=None
+    ):
+        qs, output_names = level_to_outputs(level)
         qs = torch.Tensor(qs)
-
         # Transform quantiles to homogeneus output names
         if quantiles is not None:
             _, self.output_names = quantiles_to_outputs(quantiles)
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:2038]: weird self.output_names overwritten by Super init

File: neuralforecast/losses/pytorch.py [hunk 10]
```diff
@@ -518,33 +510,56 @@ class MQLoss(torch.nn.Module):
...
-        super(MQLoss, self).__init__()
-        # Transform level to MQLoss parameters
-        qs, self.output_names = level_to_outputs(level)
+    def __init__(self, level=[80, 90], quantiles=None, horizon_weight=None):
+        qs, output_names = level_to_outputs(level)
         qs = torch.Tensor(qs)
-
         # Transform quantiles to homogeneus output names
         if quantiles is not None:
             _, self.output_names = quantiles_to_outputs(quantiles)
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:525]: weird self.output_names overwritted by Super init

File: neuralforecast/losses/pytorch.py [hunk 11]
```diff
@@ -2055,35 +2089,27 @@ def __call__(
         **Returns:**<br>
         `mqloss`: tensor (single value).
```
Comment (kdgutier) [inline neuralforecast/losses/pytorch.py:2090]: typo `mqloss`-> `hmqloss`

**Review by kdgutier — CHANGES_REQUESTED** [review]

File: neuralforecast/losses/pytorch.py (hunk 3 above)
Comment (cchallu) [reply neuralforecast/losses/pytorch.py:29]: it wont be used because it is inside other loss. Will rewrite relmse

**Review by kdgutier — APPROVED** [review]
